cupy/core/_reduction.pyx

```diff
@@ -107,6 +110,203 @@ extern "C" __global__ void ${name}(${params}) {
     return module.get_function(name)
 
 
+cpdef function.Function _create_cub_reduction_function(
+        name, block_size, items_per_thread,
+        reduce_type, params, arginfos, identity,
+        pre_map_expr, reduce_expr, post_map_expr,
+        _kernel._TypeMap type_map, preamble, options):
+
+    # move this part to a Python file for faster test cycles
+    # TODO: move it back to here!
+    from ._cub_simple_reduction import _get_cub_reduction_function_code
+    module_code = _get_cub_reduction_function_code(
+        name, block_size, items_per_thread,
+        reduce_type, params, arginfos, identity,
+        pre_map_expr, reduce_expr, post_map_expr,
+        type_map, preamble, options)
+
+    # CUB is not natively supported by NVRTC (NVlabs/cub#131), so use nvcc
+    # instead. For this, we have to explicitly spell out the default values for
+    # arch, cachd, and prepend_cupy_headers to bypass cdef/cpdef limitation...
+    # TODO(leofang): investigate Jitify for using NVRTC (also NVlabs/cub#171)
+    module = compile_with_cache(module_code, options, arch=None, cachd_dir=None,
```

Comment:
I would use some kind of context manager here to change the backend temporarily rather than to directly call the compiler in _reduction.pyx

Reply:
I am not aware of any utility we have to switch the backend in a context. `compile_with_cache` is what's used in `_reduction.pyx`. Since this is meant to be private API, I think we wanna reduce overheads as much as possible. We can revisit this in the next PR when jitify is used to support NVRTC.

A related issue is what'd happen in an HIP build. I think this feature would need to be disabled until we (I?) figure out how to use hipCUB...